bootable-jar/WFLY-14411_Bootable_Jar_Upgrade.adoc

```diff
@@ -0,0 +1,246 @@
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+We have a need to be able to upgrade the server components present in a Bootable JAR when building the jar
+from the Bootable JAR maven plugin.
+
+2 kind of components can be upgraded:
+
+* JBoss Modules module jar files (eg: undertow-core, postgresql driver that would be provisioned from third party Galleon feature-pack).
+* Galleon feature-packs that are dependencies of the Galleon feature-pack(s) in use to build a Bootable JAR.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-14411[WFLY-14411]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/WFGP-194
+
+* https://issues.redhat.com/browse/WFGP-197
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:{jkasik@redhat.com}[Jan Kasik]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [X] QE
+
+=== Affected Projects or Components
+
+* Bootable JAR Maven plugin
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [ ] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [ ] Managed domain
+
+* [ ] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+* The upgrade mechanism is done from the Bootable JAR Maven plugin.
+
+* Components that we want to upgrade are referenced with Maven Coordinates 
+(GroupId, ArtifactId, Classifier, Version and Type) from the Maven dependencies section of the Maven project.
+That is the nominal Maven way of expressing dependencies.
+
+* We are not automatically using all artifacts present in the project dependencies section for the upgrade.  
+A user needs to list the GroupId/ArtifactId/[Classifier] of the artifacts he wants to see used during the upgrade.
+
+* The `wildfly-jar-maven-plugin` configuration is evolved with an `<overridden-server-artifacts>` set that contains the artifacts that 
+are in use for the upgrade. These artifacts are upgraded globally. Any feature-pack that references it will be upgraded.
+
+* The `wildfly-jar-maven-plugin` `<FeaturePack>` configuration element is evolved with an `<overridden-artifacts>` set that contains the artifacts (JBoss Modules module 
+artifacts only) that are in use for the upgrade. These artifacts are upgraded only in the feature-pack that define them. For example we could have FP1 and FP2 that depend on
+the same GA but with different versions. In such a case we want to control which actual artifact is impacted by the upgrade. 
+
+* The overridden artifacts inside  the `<FeaturePack>` configuration element can define a version. This allows to have 
+multiple artifacts with the same GA to be upgraded to different versions according to the feature-pack that contain them.
```

Comment:
Perhaps add an example use case for this? In a chat you mentioned something about MS SQL Server.

AIUI this is a situation where FP1 ships module com.foo and FP2 ships module com.bar and both of those include artifact org.example:example.

I guess the likely scenario for this would be FP2 adds something and that something has a dep on some lib, but won't work with the version WildFly ships, so it needs to provision it's own version in a different module.

Reply:
I added the MS driver case example. The modules dependencies are identical, only the driver artifact version differs.

Reply:
@bstansberry , it appears that the driver incompatibility was not a real issue. I am removing this example, from the spec.

Reply:
I mean, the GAV are still conflicting, so the issue still exists, but having multiple drivers in the same FP would be very unlikely.

Reply:
Hmmm, perhaps it is still an valid example...I will keep it.